### PR TITLE
[multistage][bugfix] fix operator eos pull

### PR DESCRIPTION
**Pull request by walterddr**
currently agg, window, transform, fitler operator can potentially pull more than once against input for EOS.

This PR changes the behavior.
- single-input stop-the-world operator (agg, window, sort) only pull once and check for produced block first
- streaming operator (filter, transform) keeps the behavior
- dual-input, right-sided stop-the-world & left-sided stream operator (join, set) will only pull EOS on each input at most once 
    - when right side error out first, left side will not be pulled

this guarantees that no more than one EOS block will be pulled; tests are covered by existing resource queries tests

### CODE REVIEW

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java [hunk 1]
```diff
@@ -132,25 +132,22 @@ public String toExplainString() {
   @Override
   protected TransferableBlock getNextBlock() {
     try {
```
Comment (Jackie-Jiang) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:134]: Do we always need to do this try-catch? Consider moving it into base class?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/SetOperator.java [hunk 2]
```diff
@@ -89,12 +92,23 @@ public ExecutionStatistics getExecutionStatistics() {
...
-      constructRightBlockSet();
+    try {
+      if (_isTerminated) {
+        return TransferableBlockUtils.getEndOfStreamTransferableBlock();
+      }
+      if (!_isRightSetBuilt) {
+        // construct a SET with all the right side rows.
+        constructRightBlockSet();
+      }
+      if (_upstreamErrorBlock != null) {
```
Comment (Jackie-Jiang) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/SetOperator.java:103]: We probably don't need to save this as member variable. Also, the error from right table is not gathered

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java (hunk 1 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:134]: yes i can move it to base

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MultiStageOperator.java [hunk 3]
```diff
@@ -70,7 +79,7 @@ public String getOperatorId() {
   }
 
   // Make it protected because we should always call nextBlock()
-  protected abstract TransferableBlock getNextBlock();
+  protected abstract TransferableBlock getNextBlock() throws Exception;
```
Comment (Jackie-Jiang) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MultiStageOperator.java:82]: (minor) Reformat

**Review by Jackie-Jiang — APPROVED** [review]
LGTM

**Pull request merged**